File: 2022/PythonSolutions/Day08.py
from functools import reduce


def val(grid, r, c, part1):
    rgrid = list(zip(*grid))
    dirs = (grid[r][:c][::-1], grid[r][c + 1:], rgrid[c][:r][::-1], rgrid[c][r + 1:])
    if part1:
        return min(max(x or [-1]) for x in dirs) < grid[r][c]
    len_until_limit = lambda limit, it: next((i + 1 for i, x in enumerate(it) if x >= limit), len(it))
    return reduce(lambda x, y: x * y, (len_until_limit(grid[r][c], dr) for dr in dirs))


with open("../Inputs/InputDay08.txt") as file:
    m = [list(int(x) for x in line.rstrip()) for line in file]
    print(sum(val(m, r, c,  True) for r in range(len(m)) for c in range(len(m[0]))))
    print(max(val(m, r, c, False) for r in range(len(m)) for c in range(len(m[0]))))

File: 2022/rust_solutions/src/day08.rs
use std::fs;

pub fn solve() {
    let data = fs::read_to_string("../Inputs/InputDay08.txt").expect("Unable to read file");
    let data: Vec<Vec<i32>> = data.split('\n').map(
        |line| line.chars().map(|c| c.to_digit(10).unwrap() as i32).collect()
    ).collect();
    println!("part1: {}", (0..data.len()).map(|r|
            (0..data[0].len()).map(|c|
                val(&data, r, c, true)
            ).sum::<i32>()
        ).sum::<i32>()
    );
    println!("part2: {}", (0..data.len()).map(|r|
            (0..data[0].len()).map(|c|
                val(&data, r, c, false)
            ).max().unwrap()
        ).max().unwrap()
    );
}

fn val(grid: &Vec<Vec<i32>>, r: usize, c: usize, part1: bool) -> i32 {
    let dirs: Vec<Vec<i32>> = vec![
        grid[r].iter().take(c).map(|a| *a).rev().collect(),
        grid[r].iter().skip(c + 1).map(|a| *a).collect(),
        grid.iter().map(|x| x[c]).take(r).rev().collect(),
        grid.iter().map(|x| x[c]).skip(r + 1).collect(),
    ];
    if part1 {
        (dirs.iter().map(|x| *x.iter().max().unwrap_or(&-1)).min().unwrap() < grid[r][c]) as i32
    } else {
        dirs.iter().map(|dr| len_until_limit(dr, grid[r][c])).reduce(|x, y| x * y).unwrap() as i32
    }
}

fn len_until_limit(lst: &Vec<i32>, limit: i32) -> usize {
    lst.iter().enumerate().filter_map(|(i, x)|
        if *x >= limit { Some(i + 1) } else { None }
    ).next().unwrap_or(lst.len())
}
